src/functions.sh

```diff
@@ -48,6 +48,11 @@ sqm_log() { sqm_logger $VERBOSITY_INFO "$@"; }
 sqm_debug() { sqm_logger $VERBOSITY_DEBUG "$@"; }
 sqm_trace() { sqm_logger $VERBOSITY_TRACE "$@"; }
 
+
+# from https://stackoverflow.com/questions/85880/determine-if-a-function-exists-in-bash
+fn_exists() { LC_ALL=C type $1 | grep -q 'is a function'; }
```

Comment:
All the bugs and later versions aside, I am still not sure you have explained why this function is needed instead of just letting the scripts override sqm_start() like we do with everything else?

Reply:
You mean relaying on shell search order of two functions of the same name?
I am simple minded, and I believe that not having name-clashes as a designed in feature to be easier to understand. But I guess I do not fully understand what you propose.
Also I thought about moving the logging functions to their own file and conditionally source them if sqm_logger() is not a defined shell function... but that is a different PR to get to if this one actually proceeds.

Reply:
It's not shell search order, it's shell evaluation order. We source the user script at a specific place (after functions.sh). If the user script defines a function that is already defined in functions.sh, that later definition will override the one in functions.sh.

I'm pretty sure this is quite well-defined over most (all) shells, and likely to be less brittle than the fn_exists() function. I actually thought we already relied on this for something, but it appears I was mistaken; anyway, I still think it is a better approach than trying to grep the output of 'type' :)

Reply:
Also note, thanks to fn_exists sqm_start_default can warn about missing ingress() and egress() functions, at least to me that is helpful; especially since we nowhere document what a script is supposed to supply...

Reply:
"It's not shell search order, it's shell evaluation order. " okay, but that does not change the core of my argument, that sometimes explixit coding is easier to read than relying on implicir behavior. Plus the argument about warning about missing functions...

'I'm pretty sure this is quite well-defined over most (all) shells," if there is one thing that bugs me about shells it that very little can be expected to be identical between shells ;)

"I still think it is a better approach than trying to grep the output of 'type' :)" oh, sure it is an abomination that type does come with a set of posix supplied return values that encode the actual type, but we do grep games already, and this actually allows us to warn about missing functions.

Reply:
Silly idea instead of grepping type output we can grep $SCRIPT for the occurances of the function signatures, like "ingress()". Less "brittle", but also less powerful as it will not detect non-working $SCRIPT files...

Reply:
No, please don't start trying to parse the shell code of the actual script! That is worse, not better ;)

Given that you've actually gotten it to work, I guess I can live with the fn_exists wrapper for now... But I don't agree that parsing the output of 'type' is less explicit than relying on shell parsing.

As for your point about warning about missing 'ingress' and 'egress' functions, we really should supply default ones for those as well :)

Reply:
Well, not parsing really only grepping for strings like ingress(), egress() ;), but sure that is also pretty disgusting.

"As for your point about warning about missing 'ingress' and 'egress' functions, we really should supply default ones for those as well :)"

Even if we do we should also inform the users which are active to help in debugging, no?

This is topical in that the meat of a custom qos script will reside in ingress() and egress(), so silentky replacing those with versions from functions.sh might be a bit subtle, no?